File: Euclid.f90
program EuclideanGeometry
    implicit none
    
    ! 型定義：点
    type :: Point
        real :: x, y
    end type Point
    
    ! 型定義：線（2点で定義）
    type :: Line
        type(Point) :: p1, p2
    end type Line
    
    ! プロトタイプ宣言
    real :: distance
    logical :: is_parallel
    type(Point) :: translate, rotate
    real, parameter :: PI = 3.141592653589793
    
    ! テスト用変数
    type(Point) :: p1, p2, p3, p4
    type(Line) :: l1, l2
    type(Point) :: p_trans, p_rot
    real :: dist
    
    ! 点の初期化
    p1 = Point(0.0, 0.0)
    p2 = Point(1.0, 1.0)
    p3 = Point(0.0, 2.0)
    p4 = Point(1.0, 3.0)
    
    ! 線の初期化
    l1 = Line(p1, p2)
    l2 = Line(p3, p4)
    
    ! 距離計算
    dist = distance(p1, p2)
    print *, 'Distance between p1 and p2: ', dist
    
    ! 平行性判定
    if (is_parallel(l1, l2)) then
        print *, 'Lines are parallel'
    else
        print *, 'Lines are not parallel'
    end if
    
    ! 平行移動
    p_trans = translate(p1, 2.0, 3.0)
    print *, 'Translated point: (', p_trans%x, ',', p_trans%y, ')'
    
    ! 回転
    p_rot = rotate(p1, PI/2.0)
    print *, 'Rotated point: (', p_rot%x, ',', p_rot%y, ')'

contains

    ! 距離関数（体論：ユークリッド距離）
    real function distance(p, q)
        type(Point), intent(in) :: p, q
        distance = sqrt((p%x - q%x)**2 + (p%y - q%y)**2)
    end function distance
    
    ! 平行性判定（群論的構造の簡略化）
    logical function is_parallel(l1, l2)
        type(Line), intent(in) :: l1, l2
        real :: slope1, slope2
        real, parameter :: EPS = 1.0e-10
        
        ! 傾きの計算（ゼロ除算回避）
        slope1 = (l1%p2%y - l1%p1%y) / (l1%p2%x - l1%p1%x + EPS)
        slope2 = (l2%p2%y - l2%p1%y) / (l2%p2%x - l2%p1%x + EPS)
        
        ! 傾きが等しいか判定
        is_parallel = abs(slope1 - slope2) < EPS
    end function is_parallel
    
    ! 平行移動（群論：ユークリッド群の作用）
    type(Point) function translate(p, dx, dy)
        type(Point), intent(in) :: p
        real, intent(in) :: dx, dy
        translate = Point(p%x + dx, p%y + dy)
    end function translate
    
    ! 回転（群論：直交群 O(2)）
    type(Point) function rotate(p, angle)
        type(Point), intent(in) :: p
        real, intent(in) :: angle
        real :: cos_a, sin_a
        cos_a = cos(angle)
        sin_a = sin(angle)
        rotate = Point(p%x * cos_a - p%y * sin_a, p%x * sin_a + p%y * cos_a)
    end function rotate

end program EuclideanGeometry

File: Euclid.cs
using System;
using System.Collections.Generic;

// 点クラス（集合論的基盤）
public class Point
{
    public double X { get; }
    public double Y { get; }

    public Point(double x, double y)
    {
        X = x;
        Y = y;
    }

    // 距離関数（体論）
    public double DistanceTo(Point other)
    {
        return Math.Sqrt(Math.Pow(X - other.X, 2) + Math.Pow(Y - other.Y, 2));
    }

    public override string ToString() => $"({X}, {Y})";
}

// 線クラス（集合論：点の集合）
public class Line
{
    public Point P1 { get; }
    public Point P2 { get; }

    public Line(Point p1, Point p2)
    {
        if (p1.X == p2.X && p1.Y == p2.Y)
            throw new ArgumentException("Points must be distinct.");
        P1 = p1;
        P2 = p2;
    }

    // 平行性の判定（群論的変換を簡略化）
    public bool IsParallelTo(Line other)
    {
        double slope1 = (P2.Y - P1.Y) / (P2.X - P1.X + 1e-10); // ゼロ除算回避
        double slope2 = (other.P2.Y - other.P1.Y) / (other.P2.X - other.P1.X + 1e-10);
        return Math.Abs(slope1 - slope2) < 1e-10;
    }
}

// ユークリッド変換（群論：平行移動）
public class EuclideanTransformation
{
    public static Point Translate(Point p, double dx, double dy)
    {
        return new Point(p.X + dx, p.Y + dy);
    }

    // 回転（直交群 O(2) の簡略化）
    public static Point Rotate(Point p, double angleRad)
    {
        double cos = Math.Cos(angleRad);
        double sin = Math.Sin(angleRad);
        double x = p.X * cos - p.Y * sin;
        double y = p.X * sin + p.Y * cos;
        return new Point(x, y);
    }
}

// 圏論的構造の簡略化（ユークリッド空間の圏）
public class EuclideanCategory
{
    private List<Point> Points { get; } = new List<Point>();
    private List<Line> Lines { get; } = new List<Line>();

    public void AddPoint(Point p) => Points.Add(p);
    public void AddLine(Line l) => Lines.Add(l);

    // 射の例：点の変換
    public Point ApplyTransformation(Point p, Func<Point, Point> transform)
    {
        return transform(p);
    }
}

class Program
{
    static void Main()
    {
        // 点と線の生成
        Point p1 = new Point(0, 0);
        Point p2 = new Point(1, 1);
        Point p3 = new Point(0, 2);
        Point p4 = new Point(1, 3);

        Line l1 = new Line(p1, p2);
        Line l2 = new Line(p3, p4);

        // 距離の計算
        Console.WriteLine($"Distance between {p1} and {p2}: {p1.DistanceTo(p2)}");

        // 平行性の判定
        Console.WriteLine($"Are lines parallel? {l1.IsParallelTo(l2)}");

        // 変換の適用
        Point translated = EuclideanTransformation.Translate(p1, 2, 3);
        Console.WriteLine($"Translated point: {translated}");

        Point rotated = EuclideanTransformation.Rotate(p1, Math.PI / 2);
        Console.WriteLine($"Rotated point: {rotated}");

        // 圏論的構造のテスト
        EuclideanCategory category = new EuclideanCategory();
        category.AddPoint(p1);
        category.AddLine(l1);
        Point transformed = category.ApplyTransformation(p1, p => EuclideanTransformation.Rotate(p, Math.PI));
        Console.WriteLine($"Transformed point in category: {transformed}");
    }
}
